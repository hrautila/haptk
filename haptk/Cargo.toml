[package]
name = "haptk"
version = "0.3.0"
authors = ["Osma S. Rautila"]
edition = "2021"
description = "HAPTK - Haplotype analysis toolkit"
license = "GPL-3.0-only"
readme = "README.md"
homepage = "https://github.com/xosxos/haptk"
repository = "https://github.com/xosxos/haptk"
exclude = [
    "tests/*",
]


[lib]
crate-type = ["cdylib", "rlib"]
path = "src/lib.rs"
doctest = false

[[bin]]
name = "haptk"
path = "src/bin.rs"

[features]
default = ["clap"]
experimental = ["clap", "dep:fishers_exact"]
clap = ["dep:clap", "dep:anstyle"]

[dependencies]
alphanumeric-sort = "1.5"
bgzip = "0.3"
bio = "2.0"
niffler = "2.6"
ndarray = { version = "0.16", features = ["rayon", "blas", "serde"] }
ndarray-npy = { version = "0.9", default-features = false }
termion = "4.0"
color-eyre = "0.6"
rayon = "1.10"
serde = { version = "1.0", features = ["derive"] }
rand = "0.8.5"
svg = "0.13"
csv = "1.3"
itertools = "0.13"
serde_json = "1.0"
tracing = "0.1"
tracing-appender = "0.2"
indexmap = {version = "2.6", features = ["serde"]}
tracing-subscriber = { version = "0.3", default-features = false, features = ["std", "fmt", "time"] }
time = { version = "0.3", default-features = false, features = ["local-offset"] }
petgraph = { version = "0.6", default-features = false, features = ["serde-1"] }
rust-htslib = { version = "0.44", default-features = false, features = ['bzip2'] }
polars = { version = "=0.33.2", default-features = false, features = ["lazy", "is_in", "csv", "fmt", "serde"]}
image = { version = "0.25", default-features = false, features = ['png'] }
statrs = { version = "0.16", default-features = false }
chrono = { version = "0.4.38" }

# Cli
clap = { version = "=4.4.18", optional = true, features = ["derive"] }
anstyle = { version = "1.0.7", optional = true }

# Genome-wide
fishers_exact = { version="1.0.1", optional=true }


[dev-dependencies]
insta = { version = "1.26.0", default-features = false, features = ["yaml"] }



