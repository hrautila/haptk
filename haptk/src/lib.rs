// HATK - Haplotype analysis toolkit
// Copyright (C) 2022  Osma S. Rautila
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

pub mod libs;
pub use libs::{args, error, io, read_vcf, stats, structs, utils};

#[cfg(feature = "clap")]
pub use libs::clap;

/// Graphs implemented in Rust, HST graph will be replaced by the Python implementation
pub mod graphs;

/// HAPTK commands
pub mod subcommands;
