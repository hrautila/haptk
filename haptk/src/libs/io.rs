use std::collections::BTreeMap;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::{Path, PathBuf};

use color_eyre::{
    eyre::{eyre, WrapErr},
    Result,
};
use polars::prelude::*;

use crate::core::{get_csv_reader, get_input, get_tsv_reader};
use crate::structs::HapVariant;

pub fn read_variable_data_file(path: PathBuf) -> Result<DataFrame> {
    let df = CsvReader::from_path(path)?
        .with_null_values(Some(NullValues::AllColumnsSingle("NA".to_string())))
        .infer_schema(Some(500))
        .has_header(true)
        .finish()?;

    Ok(df)
}

#[derive(Debug, Clone, serde::Deserialize)]
struct RecombinationRow<'a> {
    _chr: &'a str,
    pos: u64,
    _rate: f32,
    cm: f32,
}

pub fn read_recombination_file(path: PathBuf) -> Result<BTreeMap<u64, f32>> {
    let mut rdr = get_tsv_reader(get_input(Some(path))?, false);
    let mut rates = BTreeMap::new();

    let mut last_cm = 0.0;
    let mut last_pos = 0;
    for line in rdr.records() {
        let record = line?;
        let row: RecombinationRow = record.deserialize(None).
            wrap_err(eyre!("Make sure no headers are present and that the recombination file is in order chr,pos,rate,cm. If the issue is not fixed, you have an invalid field in the file."))?;
        if last_cm > row.cm {
            return Err(eyre!(
                "Recombination rates file is not sorted for centimorgans"
            ));
        }
        if last_pos > row.pos {
            return Err(eyre!(
                "Recombination rates file is not sorted for positions"
            ));
        }
        last_cm = row.cm;
        last_pos = row.pos;
        rates.insert(row.pos, row.cm);
    }

    Ok(rates)
}

pub fn read_haplotype_file(ht_path: PathBuf) -> Result<Vec<HapVariant>> {
    let input = get_input(Some(ht_path))?;
    let mut rdr = get_csv_reader(input);

    let mut variants: Vec<HapVariant> = vec![];

    for line in rdr.records() {
        let record = line?;
        let variant: HapVariant = record.deserialize(None)?;
        if let Some(latest) = variants.last() {
            if latest.pos > variant.pos {
                return Err(eyre!(
                    "The haplotype file is not sorted by position. {} is larger than {}",
                    latest,
                    variant
                ));
            }
        }
        variants.push(variant);
    }

    Ok(variants)
}

pub fn read_lines<P>(filename: P) -> Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    // let file = File::open(filename)?;
    let name = filename.as_ref().display();
    let file = match File::open(&filename) {
        Ok(x) => x,
        Err(err) => {
            let msg = format!("failed to open {name}: {err}");
            return Err(std::io::Error::new(std::io::ErrorKind::NotFound, msg))?;
        }
    };
    Ok(io::BufReader::new(file).lines())
}

pub fn read_sample_ids(path: &Option<PathBuf>) -> Result<Option<Vec<String>>> {
    match path {
        Some(path) => {
            let mut samples = vec![];

            for line in read_lines(path)?.flatten() {
                let line = line.trim();
                samples.push(line.to_string());
            }
            Ok(Some(samples))
        }
        None => Ok(None),
    }
}
