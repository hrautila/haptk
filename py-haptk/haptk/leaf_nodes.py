import itertools

def find_leaf_neighbors(hst):
    # A set of sibling nodes from the tree
    neighbor_sets = []

    for idx in hst.node_indices():
        # A node is leaf if it has no outgoing edges
        if len(hst.out_edges(idx)) == 0:

            # Checks if all siblings have exactly 1 sample
            is_paired_leaf = True

            # Add the first sibling
            neighbors = [idx]

            # Get parents by moving up to adj_direction
            parents = hst.adj_direction(idx, True)

            # Iterate parent for siblings (I think this loops only once in HSTs)
            for parent in parents:
                for sibling in hst.adj_direction(parent, False):
                    # Dont re-add the first sibling
                    if sibling != idx:
                        data = hst.get_node_data(sibling)

                        # Dont add if there is more than 1 sample in the node
                        if len(data['indexes']) == 1:
                            neighbors.append(sibling)
                            pass
                        else:
                            # A sibling has more than 1 sample
                            is_paired_leaf = False

            if is_paired_leaf:
                neighbors.sort()
                neighbor_sets.append(neighbors)

                
    # Sort
    neighbor_sets.sort()

    # Dedup because looping all leaf nodes produces duplicate pairs
    neighbor_sets = list(k for k,_ in itertools.groupby(neighbor_sets))

    return neighbor_sets

def find_average_leaf_node_length(hst):
    leaf_nodes = get_leaf_nodes(hst)

    sum = 0

    for node_idx in leaf_nodes:
        node_data = hst.get_node_data(node_idx)

        # A saturating substraction just to be sure
        length = max(node_data['stop'] - node_data['start'], 0)
        sum += length

    return sum / len(leaf_nodes)

def get_leaf_nodes(hst):
    leaf_nodes = []

    for idx in hst.node_indices():
        if len(hst.out_edges(idx)) == 0:
            leaf_nodes.append(idx)

    return leaf_nodes

   
