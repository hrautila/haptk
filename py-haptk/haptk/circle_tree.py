from ete3 import TreeStyle, TextFace, CircleFace,  NodeStyle, add_face_to_node

from haptk import utils

def circle_tree_style():
    ts = TreeStyle()

    ts.branch_vertical_margin = 10
    ts.show_leaf_name = False
    ts.show_scale = False

    ts.mode = "c"
    ts.arc_start = 0 # 0 degrees = 15:00 o'clock
    ts.arc_span = 355
    ts.min_leaf_separation = 0
    ts.root_opening_factor = 0
    ts.optimal_scale_level = "mid"
    ts.allow_face_overlap = True

    return ts

def is_maj_up_to(hst, node, value):
    iter_node = node
    # print(f"checking {iter_node.name} for majority")
    for _ in range(0, value):
        if iter_node.name == "0":
            return True

        if not iter_node.up.name == "0":
            # print("input node:", iter_node.up)
            is_big = is_majority_node(hst, iter_node.up)
            if is_big == False:
                # print(f"run {_}: {iter_node.up.name} is not big")
                return False

        iter_node = iter_node.up
            
    return True

def is_majority_node(hst, node):
    parent = node.up
    input_node_data = hst.get_node_data(int(node.name))
    value = len(input_node_data["indexes"])
    # print(f"input node {node.name} has {value} indexes")

    siblings = {}
    for child in parent.children:
        data = hst.get_node_data(int(child.name))
        if not child.name == node.name:
            siblings[child.name] = len(data["indexes"])


    # print(value, siblings.values())
    for v in siblings.values():
        if value < v: 
            return False
  

def get_children_idx_amounts(hst, node):
    return [len(hst.get_node_data(c.name)["indexes"]) for c in node.children]

def draw_circle_tree(hst, t, output, samples, samples_to_tag, colors):
    style = NodeStyle()
    style["size"] = 0

    narea_color = "#DAF7A6"
    larea_color = "#FFB6C1"

    spent_parents = []
    spent_grand_parents = []

    for n in t.traverse():
        n.set_style(style)
        
        if not n.is_leaf():
            indexes = hst.get_node_indexes(n.name)
            label = len(indexes)
            F = TextFace(label, tight_text=True, penwidth=30, fsize=8)
            F.rotation = 270
            n.add_face(F, column=0)

            count = 0
            for i in get_children_idx_amounts(hst, n):
                if i > 80:
                    count += 1

            if count > 1:
                haplotype = hst.get_node_haplotype(n.name)
                F = TextFace(f"{len(haplotype)}", fsize=30, tight_text=True, fgcolor="red")
                # F.rotation = 270
                n.add_face(F, column=0)
                
                

        if n.is_leaf():
            if samples_to_tag == []:
                n.set_style(utils.return_node_style("#000", 0))
                n.img_style["bgcolor"] = "#FFF"

            for (color_i, sample_list) in enumerate(samples_to_tag):
                results = [i for i in indexes if samples[i] in sample_list]
                if results:
                    n.set_style(utils.return_node_style(colors[color_i], 0))
                    n.img_style["bgcolor"] = colors[color_i]

            if n.up:
                parent = n.up
                parent_indexes = hst.get_node_indexes(parent.name)

                if len(parent_indexes) == 2:
                    if parent in spent_parents:
                        if is_maj_up_to(hst, n, 14):
                            # print(f"{n.name} is_maj")

                            haplotype = hst.get_node_haplotype(n.name)
                            start_idx = hst.get_node_start_idx(n.name)
                            stop_idx = hst.get_node_stop_idx(n.name)
                            start = hst.coords[start_idx]
                            stop = hst.coords[stop_idx]
                            # print(start, stop)
                            bp = stop['pos'] - start['pos']
                            F = TextFace(f"{len(haplotype)}, {bp}", fsize=30, fgcolor="white")
                            # F.margin_left = 50
                            n.add_face(F, column=0)

                            # n.set_style(utils.return_node_style("white", 0))
                            # n.img_style["bgcolor"] = "white"
                    else:
                        spent_parents.append(parent)


    return t

