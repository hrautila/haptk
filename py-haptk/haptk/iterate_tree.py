from ete3 import TreeStyle, TextFace, CircleFace,  NodeStyle, add_face_to_node
from scipy.stats import fisher_exact

from haptk import utils

def iterate_tree_style():
    ts = TreeStyle()

    ts.branch_vertical_margin = 10
    ts.show_leaf_name = False
    ts.show_scale = False

    ts.mode = "c"
    ts.arc_start = 0 # 0 degrees = 15:00 o'clock
    ts.arc_span = 355
    ts.min_leaf_separation = 0
    ts.root_opening_factor = 0
    ts.optimal_scale_level = "mid"
    ts.allow_face_overlap = True

    return ts

def iterate_tree_inner(hst, t, samples_to_tag):
    style = NodeStyle()
    style["size"] = 0

    for n in t.traverse():
        n.set_style(style)
        
        indexes = hst.get_node_indexes(n.name)

        tot_cases, tot_ctrls = 0, 0
        n_cases, n_ctrls = [], []

        for (i, sample_list) in enumerate(samples_to_tag):
            if i == 0:
                tot_cases = len(sample_list)
                n_cases = [i for i in indexes if hst.samples[i] in sample_list]
            if i == 1:
                tot_ctrls = len(sample_list)
                n_ctrls = [i for i in indexes if hst.samples[i] in sample_list]

        n_cases = len(n_cases)
        n_ctrls = len(n_ctrls)
        # res = fisher_exact([[n_cases, n_ctrls], [tot_cases - n_cases, tot_ctrls - n_ctrls]], alternative="greater")
        res = fisher_exact([[n_cases, n_ctrls], [tot_cases - n_cases, tot_ctrls - n_ctrls]])
        
        print(f"{n.name},{res.pvalue},{n_cases},{n_ctrls},{tot_cases},{tot_ctrls}")
        if res.pvalue < 0.0005:
            label = len(indexes)
            F = TextFace(f"{res.pvalue:.4}", tight_text=True, penwidth=30, fsize=8, fgcolor='red')
            F.rotation = 270
            F.background.color = "#FFF"
            F.margin_right = 10
            n.add_face(F, column=0)

        # if n.is_leaf() and res.pvalue < 0.005:
        label = len(indexes)
        F = TextFace(label, tight_text=True, penwidth=30, fsize=8)
        F.rotation = 270

        mix = res.pvalue * 1000
        if mix > 1:
            mix = 1
        color = utils.color_fader('red', 'blue', mix)

        F.background.color = color
        n.add_face(F, column=0)

    return t
