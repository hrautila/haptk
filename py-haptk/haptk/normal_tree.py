from ete3 import TreeStyle, TextFace, NodeStyle

from haptk import utils

def normal_tree_style():
    ts = TreeStyle()
    ts.rotation = 90
    ts.show_leaf_name = False
    ts.show_scale = False
    ts.min_leaf_separation = 0
    ts.branch_vertical_margin = 5
    ts.optimal_scale_level = "full"
    ts.allow_face_overlap = True
    return ts


def draw_normal_tree(hst, t, samples_to_tag, colors, proportions, show_haplotype, n_markers, show_pos):
    style = NodeStyle()
    style["size"] = 0

    for n in t.traverse():
        n.set_style(style)

        node_data = hst.get_node_data(int(n.name))

        label = len(node_data["indexes"])

        if proportions:
            top_data = hst.get_node_data(int('0'))
            samplen = len(top_data["indexes"])
            prc = label / samplen
            label = f"N={label} ({prc:.2f})"
        else:
            label = f"{label}"

        if show_haplotype:
            label = f"N={len(node_data['indexes'])}\n{hst.haplotype_string(node_data, n_markers)}"

        if show_pos:
            label = f"N={len(node_data['indexes'])}"
            haplotype = hst.haplotype_string(node_data, n_markers)
            start = node_data['start']['pos']
            stop = node_data['stop']['pos']
            label = f"{label}\n{haplotype}\nstart: {start}\nstop: {stop}"

        F = TextFace(label, tight_text=True, penwidth=30)
        F.rotation = 270
        n.add_face(F, column=0)

        if n.is_leaf():
            if samples_to_tag == []:
                n.set_style(utils.return_node_style("#000", 1))

            for (color_i, sample_list) in enumerate(samples_to_tag):
                results = [i for i in node_data["indexes"] if hst.get_sample_name(i) in sample_list]
                if results:
                    n.set_style(utils.return_node_style(colors[color_i], 7 - color_i))


    return t

