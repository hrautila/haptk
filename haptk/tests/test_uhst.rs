mod common;

#[cfg(test)]
#[cfg(feature = "clap")]
mod test_uhst {
    use super::*;

    use color_eyre::Result;

    use std::io::Write;
    use std::path::PathBuf;

    use haptk::{
        args::{Selection, StandardArgs},
        read_vcf::read_vcf_to_matrix,
        structs::Coord,
        subcommands::uhst_shard::{self, LocDirection},
    };

    #[test]
    fn matrix_to_graph() -> Result<()> {
        let args = StandardArgs {
            file: PathBuf::from(common::TEST_VCF),
            ..Default::default()
        };
        let mut vcf = read_vcf_to_matrix(&args, "chr9", 32, None, None, false).unwrap();

        let coord = Coord {
            contig: String::from("chr9"),
            reference: String::from("G"),
            alt: String::from("T"),
            pos: 32,
        };

        let g = uhst_shard::construct_uhst(&mut vcf, &LocDirection::Left, &coord, 1, false)?;
        let mut f = std::fs::File::create("tests/results/test.dot").unwrap();
        f.write_all(format!("{}", petgraph::dot::Dot::new(&g)).as_bytes())
            .unwrap();
        Ok(())
    }

    //--- Binary tests

    #[test]
    fn uhst_all() {
        run_uhst(Selection::All);

        let res = std::fs::read_to_string("tests/results/uhst_mbah.csv").unwrap();
        insta::assert_yaml_snapshot!(res);

        let res = std::fs::read_to_string("tests/results/uhst_shared_core_haplotype.csv").unwrap();
        insta::assert_yaml_snapshot!(res);
    }

    #[test]
    fn uhst_only_ref() {
        run_uhst(Selection::OnlyRefs);

        let res = std::fs::read_to_string("tests/results/uhst_mbah_only_refs.csv").unwrap();
        insta::assert_yaml_snapshot!(res);

        let res = std::fs::read_to_string("tests/results/uhst_shared_core_haplotype_only_refs.csv")
            .unwrap();
        insta::assert_yaml_snapshot!(res);
    }

    #[test]
    fn uhst_only_alt() {
        run_uhst(Selection::OnlyAlts);

        let res = std::fs::read_to_string("tests/results/uhst_mbah_only_alts.csv").unwrap();
        insta::assert_yaml_snapshot!(res);

        let res = std::fs::read_to_string("tests/results/uhst_shared_core_haplotype_only_alts.csv")
            .unwrap();
        insta::assert_yaml_snapshot!(res);
    }

    #[test]
    fn uhst_only_longest() {
        run_uhst(Selection::OnlyLongest);

        let res = std::fs::read_to_string("tests/results/uhst_mbah_only_longest.csv").unwrap();
        insta::assert_yaml_snapshot!(res);

        let res =
            std::fs::read_to_string("tests/results/uhst_shared_core_haplotype_only_longest.csv")
                .unwrap();
        insta::assert_yaml_snapshot!(res);
    }

    fn run_uhst(selection: Selection) {
        let args = common::clap_standard_args(selection);

        let cmd = haptk::clap::SubCommand::Uhst {
            args,
            log_and_verbosity: crate::common::silent_verbosity(),
            threads: 8,
            min_size: 1,
            publish: false,
            sharded: false,
        };
        haptk::clap::run_cmd(cmd).unwrap();
    }
}
