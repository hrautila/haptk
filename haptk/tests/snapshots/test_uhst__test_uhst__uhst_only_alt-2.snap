---
source: haptk/tests/test_uhst.rs
expression: res
---
"<svg style=\"background-color:white\" viewBox=\"0 0 20000 14000\" xmlns=\"http://www.w3.org/2000/svg\">\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"10000\" y=\"280\">\n15\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"9333.333\" y=\"532\">\n14\n</text>\n<path d=\"M10210,317.8 L10210,490 L9595.833,490\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"8666.666\" y=\"784\">\n13\n</text>\n<path d=\"M9543.333,569.8 L9543.333,742 L8929.166,742\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"7999.9995\" y=\"1036\">\n12\n</text>\n<path d=\"M8876.666,821.8 L8876.666,994 L8262.5,994\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"7333.333\" y=\"1288\">\n11\n</text>\n<path d=\"M8210,1073.8 L8210,1246 L7595.833,1246\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"6666.6665\" y=\"1540\">\n10\n</text>\n<path d=\"M7543.333,1325.8 L7543.333,1498 L6929.1665,1498\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"5999.9995\" y=\"1792\">\n9\n</text>\n<path d=\"M6876.6665,1577.8 L6876.6665,1750 L6146.9995,1750\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"5333.333\" y=\"2044\">\n8\n</text>\n<path d=\"M6062.9995,1829.8 L6062.9995,2002 L5480.333,2002\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"4666.6665\" y=\"2296\">\n7\n</text>\n<path d=\"M5396.333,2081.8 L5396.333,2254 L4813.6665,2254\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"4000\" y=\"2548\">\n6\n</text>\n<path d=\"M4729.6665,2333.8 L4729.6665,2506 L4147,2506\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"3333.3333\" y=\"2800\">\n5\n</text>\n<path d=\"M4063,2585.8 L4063,2758 L3480.3333,2758\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"2666.6667\" y=\"3052\">\n4\n</text>\n<path d=\"M3396.3333,2837.8 L3396.3333,3010 L2813.6667,3010\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"2000\" y=\"3304\">\n3\n</text>\n<path d=\"M2729.6667,3089.8 L2729.6667,3262 L2147,3262\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"1333.3334\" y=\"3556\">\n2\n</text>\n<path d=\"M2063,3341.8 L2063,3514 L1480.3334,3514\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"666.6667\" y=\"3808\">\n1\n</text>\n<path d=\"M1396.3334,3593.8 L1396.3334,3766 L813.6667,3766\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"666.6667\" y=\"3808\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"2000\" y=\"3808\">\n1\n</text>\n<path d=\"M1396.3334,3593.8 L1396.3334,3766 L1979,3766\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"2000\" y=\"3808\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"3333.3335\" y=\"3556\">\n1\n</text>\n<path d=\"M2063,3341.8 L2063,3514 L3312.3335,3514\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"3333.3335\" y=\"3556\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"4666.667\" y=\"3304\">\n1\n</text>\n<path d=\"M2729.6667,3089.8 L2729.6667,3262 L4645.667,3262\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"4666.667\" y=\"3304\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"6000\" y=\"3052\">\n1\n</text>\n<path d=\"M3396.3333,2837.8 L3396.3333,3010 L5979,3010\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"6000\" y=\"3052\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"7333.333\" y=\"2800\">\n1\n</text>\n<path d=\"M4063,2585.8 L4063,2758 L7312.333,2758\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"7333.333\" y=\"2800\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"8666.666\" y=\"2548\">\n1\n</text>\n<path d=\"M4729.6665,2333.8 L4729.6665,2506 L8645.666,2506\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"8666.666\" y=\"2548\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"10000\" y=\"2296\">\n1\n</text>\n<path d=\"M5396.333,2081.8 L5396.333,2254 L9979,2254\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"10000\" y=\"2296\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"11333.332\" y=\"2044\">\n1\n</text>\n<path d=\"M6062.9995,1829.8 L6062.9995,2002 L11312.332,2002\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"11333.332\" y=\"2044\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"12666.666\" y=\"1792\">\n1\n</text>\n<path d=\"M6876.6665,1577.8 L6876.6665,1750 L12645.666,1750\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"12666.666\" y=\"1792\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"14000\" y=\"1540\">\n1\n</text>\n<path d=\"M7543.333,1325.8 L7543.333,1498 L13979,1498\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"14000\" y=\"1540\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"15333.332\" y=\"1288\">\n1\n</text>\n<path d=\"M8210,1073.8 L8210,1246 L15312.332,1246\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"15333.332\" y=\"1288\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"16666.666\" y=\"1036\">\n1\n</text>\n<path d=\"M8876.666,821.8 L8876.666,994 L16645.666,994\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"16666.666\" y=\"1036\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"18000\" y=\"784\">\n1\n</text>\n<path d=\"M9543.333,569.8 L9543.333,742 L17979,742\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"18000\" y=\"784\">\n1\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"19333.332\" y=\"532\">\n1\n</text>\n<path d=\"M10210,317.8 L10210,490 L19312.332,490\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"19333.332\" y=\"532\">\n1\n</text>\n</svg>"

